### PR TITLE
client/setec: allow a Store to serve secrets entirely locally

### PR DESCRIPTION
**Pull request by creachadair**
Add a StoreClient interface that implements the parts of the setec.Client that
the Store uses. Add a FileClient type that implements this interface, and
modify the Store to accept he interface instead of a concrete Client.

A FileClient uses a local JSON file containing stored secrets, and serves them
locally without using the network. This permits a program using a Store to
start up entirely from local storage, provided the required secrets are saved
locally, even if no setec service is available.

This differs from the cache in that a cache is intended for the case where you
do have a server, and want to work around it being temporarily unavailable.  A
FileClient, by contrast, is intended for bootstrapping or in deployments where
a secrets server is never available.

The operator can construct a secrets file by hand, or by copying the output
written by a FileCache (the file formats are compatible).

### CODE REVIEW

File: client/setec/fileclient.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+//	   "secret-name-1": {
+//	      "secret": {"Value": "b3BlbiBzZXNhbWU=", "Version": 1}
+//	   },
+//	   "secret-name-2": {
+//	      "secret": {"Value": "eHl6enk=", "Version": 5}
+//	   },
+//	   ...
+//	}
+//
+// The secret values are encoded as base64 strings. A cache file written out by
```
Comment (creachadair) [inline client/setec/fileclient.go:39]: I have not yet, but could easily add a subcommand to the CLI to generate these files.

File: client/setec/client_test.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+const testSecrets = `{
+  "apple": {
+     "secret": {"Version": 1, "Value": "c2F1Y2U="}
+  },
+  "pear": {
+     "unrelated": true
+  },
+  "plum": {
+     "secret": {"Version": 2, "Value": "dGFydGxldA=="}
```
Comment (danderson) [inline client/setec/client_test.go:25]: UX suggestion: maybe this struct could have both a Value and a TextValue field, where the latter is defined to be Value without base64 encoding?

We do still need the B64 version for generality with binary secrets, but having a second field where plaintext secrets can be provided verbatim would make this much more usable for 90% of needs without additional tooling.

Fine as a followup though, and maybe not worth the new confusing precedence semantics and so forth?...

Comment (danderson) [reply client/setec/client_test.go:25]: Another schema option: a `Format` field, with valid values `plain` and `base64`, and "" implying `base64` for backwards compatibility? A fair bit more typing, but maybe better by being more explicit? :shrug:

File: client/setec/client_test.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  },
+  "cherry": {
```
Comment (danderson) [inline client/setec/client_test.go:27]: Add a test for `foo/bar/baz` style secrets? We don't treat `/` specially, but it's a convention we're encouraging so would be good to lock it in with a test case if that changes someday.

**Review by danderson — APPROVED** [review]

File: client/setec/client_test.go (hunk 2 above)
Comment (creachadair) [reply client/setec/client_test.go:25]: Good ideas. I think I have a mild preference for the first (Value/TextValue), but either would be fine if you prefer the latter.

File: client/setec/client_test.go (hunk 3 above)
Comment (creachadair) [reply client/setec/client_test.go]: Good call, done.

File: client/setec/client_test.go (hunk 2 above)
Comment (creachadair) [reply client/setec/client_test.go:25]: Updated, and some additional tests. PTAL.

File: client/setec/fileclient.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	for name, val := range input {
+		if name == "" || val.Secret == nil {
+			continue // no secret value; skip
+		}
+		sec := val.Secret
+		if sec.Version <= 0 || (sec.TextValue == "" && len(sec.Value) == 0) {
+			continue // invalid version or no value
+		}
+		if sec.TextValue != "" {
+			db[name] = &api.SecretValue{Value: []byte(sec.TextValue), Version: sec.Version}
```
Comment (danderson) [inline client/setec/fileclient.go:68]: hmm, shouldn't one of these two branches be either B64 decoding or encoding?

... Oh, no, duh, []byte's json marshaling is what does the B64, once we're in this layer everything's already decoded.

**Review by danderson — APPROVED** [review]

**Pull request merged**